pkg/controllers/workgenerator/controller.go

```diff
@@ -874,19 +876,30 @@ func (r *Reconciler) SetupWithManager(mgr controllerruntime.Manager) error {
 						"Failed to process an update event for work object")
 					return
 				}
-				oldAppliedStatus := meta.FindStatusCondition(oldWork.Status.Conditions, fleetv1beta1.WorkConditionTypeApplied)
-				newAppliedStatus := meta.FindStatusCondition(newWork.Status.Conditions, fleetv1beta1.WorkConditionTypeApplied)
-				oldAvailableStatus := meta.FindStatusCondition(oldWork.Status.Conditions, fleetv1beta1.WorkConditionTypeAvailable)
-				newAvailableStatus := meta.FindStatusCondition(newWork.Status.Conditions, fleetv1beta1.WorkConditionTypeAvailable)
-
-				// we only need to handle the case the applied or available condition is changed between the
-				// new and old work objects. Otherwise, it won't affect the binding applied condition
-				if condition.EqualCondition(oldAppliedStatus, newAppliedStatus) && condition.EqualCondition(oldAvailableStatus, newAvailableStatus) {
-					klog.V(2).InfoS("The work applied or available condition didn't flip between true and false, no need to reconcile", "oldWork", klog.KObj(oldWork), "newWork", klog.KObj(newWork))
-					return
+				oldAppliedCondition := meta.FindStatusCondition(oldWork.Status.Conditions, fleetv1beta1.WorkConditionTypeApplied)
+				newAppliedCondition := meta.FindStatusCondition(newWork.Status.Conditions, fleetv1beta1.WorkConditionTypeApplied)
+				oldAvailableCondition := meta.FindStatusCondition(oldWork.Status.Conditions, fleetv1beta1.WorkConditionTypeAvailable)
+				newAvailableCondition := meta.FindStatusCondition(newWork.Status.Conditions, fleetv1beta1.WorkConditionTypeAvailable)
+
+				// we try to filter out events, we only need to handle the updated event if the applied or available condition flip between true and false
+				// or the failed placements are changed.
+				if condition.EqualCondition(oldAppliedCondition, newAppliedCondition) && condition.EqualCondition(oldAvailableCondition, newAvailableCondition) {
+					if !condition.IsConditionStatusTrue(newAppliedCondition, newWork.Generation) || !condition.IsConditionStatusTrue(newAvailableCondition, newWork.Generation) {
+						// we need to compare the failed placement if the work is not applied or available
+						oldFailedPlacements := extractFailedResourcePlacementsFromWork(oldWork)
+						newFailedPlacements := extractFailedResourcePlacementsFromWork(newWork)
+						if cmp.Equal(oldFailedPlacements, newFailedPlacements, cmp.Options{cmpopts.SortSlices(condition.LessFuncResourceIdentifier),
```

Comment:
we cannot use the "cmp" package and this package is only used for testing purpose. :(

You can use "equality.Semantic.DeepEqual" instead. If i remember it correctly, it will ignore the slice order.

Reply:
It is intended to only be used in tests, as performance is not a goal and it may panic if it cannot compare the values. Its propensity towards panicking means that its unsuitable for production environments where a spurious panic may be fatal.

https://pkg.go.dev/github.com/google/go-cmp/cmp